karoo_gp/engine.py

```diff
@@ -37,28 +38,94 @@ def _type_check(self, trees):
         if not all(isinstance(tree, Tree) for tree in trees):
             raise TypeError(f"trees must be a sequence of Trees")
 
+    def parse_expr(self, expr, X_dict, shape):
+        """Parse an expr into a function and insert sample X_dict"""
+        tree = ast.parse(expr, mode='eval').body
+        return self.parse_node(tree, X_dict, shape)
+
+    def parse_node(self, node, X_dict, shape):
+        """Recursively build a function from ast tree and sample X_dict"""
+        if isinstance(node, ast.IfExp):
+            return self.operators['if'](
+                self.parse_node(node.test, X_dict, shape),
+                self.parse_node(node.body, X_dict, shape),
+                self.parse_node(node.orelse, X_dict, shape))
+        if isinstance(node, ast.Name):
+            return X_dict[node.id]
+        elif isinstance(node, ast.Num):
+            return self.num_obj(shape, node.n)
+        elif isinstance(node, ast.UnaryOp):
+            return self.operators[type(node.op)](
+                self.parse_node(node.operand, X_dict, shape))
+        elif isinstance(node, ast.BinOp):
+            return self.operators[type(node.op)](
+                self.parse_node(node.left, X_dict, shape),
+                self.parse_node(node.right, X_dict, shape))
+        elif isinstance(node, ast.BoolOp):
+            return self.operators[type(node.op)](
+                *[self.parse_node(v, X_dict, shape) for v in node.values])
+        elif isinstance(node, ast.Compare):
+            return self.operators[type(node.ops[0])](
+                self.parse_node(node.left, X_dict, shape),
+                self.parse_node(node.comparators[0], X_dict, shape))
+        elif isinstance(node, ast.Call):
+            return self.operators[node.func.id.lower()](
+                self.parse_node(node.args[0], X_dict, shape))
+
+
 #++++++++++++++++++++++++++++
 #   Numpy                   |
 #++++++++++++++++++++++++++++
 
-def inf_to_zero_divide(a, b):
-    # This continued to raise 'RuntimeWarning: divide by zero' errors. This
-    # is a proposed solution. ref: https://stackoverflow.com/a/64747978
-    with np.errstate(divide='ignore'):
-        out = np.where(b==0, 0., a / b)
-    return out
+def safe_divide(a, b):
+    """If dividing by 0, return 0"""
+    return np.where(b==0, 0, a/b)
```

Comment:
Ah, I assumed the function was working with regular `int`s/`float`s, but instead it works with numpy arrays (maybe it should be clarified in the docstring):
```pycon
>>> import numpy as np
>>> a = np.array([1, 2, 3])
>>> b = np.array([2, 1, 0])
>>> np.where(b == 0, 0, a / b)
<stdin>:1: RuntimeWarning: divide by zero encountered in true_divide
array([0.5, 2. , 0. ])
```
Now I understand how it works, but there are a few weird behaviors involved:
* dividing `int`s/`float`s by `0` raises an error, whereas with numpy `array`s it's just a warning
* `a/b` is still evaluated before the call to `np.where`, but since it's now a warning the execution continues
* since it's evaluated before the call, `np.where` can't do anything to catch/silence errors or warnings, in fact the warning is still raised (see my example above)
* in your example above the warnings is not raised not because `np.where` silences it, but because by default Python hides repeated warnings.  In fact, if you try with `python3 -Walways` it will show up both times.

As for how `np.where` actually works, it seems that both `b == 0` and `a / b` return array:
```pycon
>>> b == 0
array([False, False,  True])
>>> a / b
array([0.5, 2. , inf])
```
`np.where` just goes through the two arrays in parallel and replaces items in the latter with the specified value (`0`) whenever there is a `True` in the former.

Reply:
How about this? It silences the warnings (both `divide` and `invalid` kwargs required):
```
def safe_divide(a, b):
    """If dividing by 0, return 0"""
    with np.errstate(divide='ignore', invalid='ignore'):
        output = np.where(b==0, 0, a/b)
    return output
```

Reply:
If that silences the warnings (you could clarify that in a comment), then it's ok.

Reply:
I found something better, doesn't seem to throw any errors:
```
  nonzero = b != 0
  c = np.zeros(a.shape)
  c[nonzero] = a[nonzero] / b[nonzero]
```